File: pacman-reloaded/src/main/java/map/color_parser.rb
require 'java'

java_import Java::resource.Resource
java_import Java::javax.imageio.ImageIO
java_import Java::java.awt.image.BufferedImage
java_import Java::com.uqbar.vainilla.appearances.Sprite
java_import Java::java.awt.Color

class ColorParser
  attr_accessor :walking_matrix, :background, :block_size

  COLOR_HASH = {
      -324853 => Color::RED,
      -1 => Color::BLUE,
      -301813 => Color::ORANGE,
      -16046085 => Color::LIGHT_GRAY,
      -14550261 => Color::BLACK,
      -787701 => Color::PINK,
      -324638 => Color::GREEN,
      -277237 => Color::YELLOW
  }

  def initialize(block_size)
    self.block_size = block_size
    self.walking_matrix = []
  end

  def parse(file)
    create_walking_matrix(file)
    create_background(file)
  end

  def create_background(file)
    read_image(file) do |image,width, height|
      new_image = BufferedImage.new width * block_size, height * block_size, BufferedImage::TYPE_INT_ARGB
      graphics = new_image.createGraphics
      (0...height).each do |row_number|
        create_image_row(row_number, width, image, graphics)
      end
      graphics.dispose
      self.background = Sprite.new new_image
    end
  end

  def create_walking_matrix(file)
    read_image(file) do |image,w, h|
      (0...h).each do |i|
        create_row(i, w, image)
      end
    end
  end

  def read_image(file)
    stream = Resource.getResourceAsStream file
    image = ImageIO.read stream
    w = image.width
    h = image.height
    yield image, w, h
  end

  private

  def create_image_row(row_number, width, image, graphics)
    (0...width).each do |column_number|
      pixel = image.getRGB(column_number, row_number)
      g_image = image_for(pixel, image)
      graphics.draw_image(g_image, nil, column_number* block_size, row_number * block_size)
    end
  end

  def create_row(i, width, image)
    row = []
    (0...width).each do |j|
      pixel = image.getRGB(j, i)
      row << boolean_for(pixel)
    end
    add_row row
  end

  def add_row(row)
    walking_matrix << row
  end


  def boolean_for(color)
    color != -324853
  end

  def image_for(color, image)
    new_image = BufferedImage.new block_size, block_size, BufferedImage::TYPE_INT_ARGB
    graphics = new_image.createGraphics
    puts color
    kolor = COLOR_HASH[color]
    if kolor.blank?
      raise "No color found for #{color}"
    end
    graphics.color = kolor
    graphics.fill_rect( 0, 0, block_size, block_size)
    graphics.dispose
    new_image
  end
end

File: pacman-reloaded/src/main/java/map/color/ColorParser.java
package map.color;


import com.uqbar.vainilla.appearances.Sprite;
import resource.Resource;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ColorParser {

    public static Map<Integer, Color> colorMap = new HashMap<Integer, Color>();

    static {
        colorMap.put(-324853, Color.RED);
        colorMap.put(-1, Color.BLUE);
        colorMap.put(-301813, Color.ORANGE);
        colorMap.put(-16046085 , Color.LIGHT_GRAY);
        colorMap.put(-14550261, Color.BLACK);
        colorMap.put(-787701, Color.PINK);
        colorMap.put(-277237, Color.YELLOW);
    }

    private int blockSize;
    private Sprite background;
    private List<List<Boolean>> walkingMatrix;

    public ColorParser(int blockSize){
        setBlockSize(blockSize);
        setWalkingMatrix(new ArrayList<List<Boolean>>());

    }

    public void parse(String file){
        createBackground(file);
        create_walking_matrix(file);
        //self.background
    }

    private void create_walking_matrix(String file) {
        InputStream stream = Resource.class.getResourceAsStream(file);
        BufferedImage image;
        try {
            image = ImageIO.read(stream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        int i = image.getWidth();
        int h = image.getHeight();

        for (int j = 0; j <h; j++) {
            create_row(j, i, image);
        }
        try {
            stream.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void create_row(int i,int  width,BufferedImage image){
        List<Boolean>row = new ArrayList<Boolean>();
        for (int j = 0; j < width; j++) {
            int pixel = image.getRGB(j, i);
            row.add(booleanFor(pixel));
        }
        addRow(row);
    }

    public void  addRow(List<Boolean>row){
        getWalkingMatrix().add(row);
    }

    public boolean booleanFor(int color){
     return color != -324853;
    }

    public void createBackground(String file) {
        InputStream stream = Resource.class.getResourceAsStream(file);
        BufferedImage image;
        try {
            image = ImageIO.read(stream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        int i = image.getWidth();
        int h = image.getHeight();



        BufferedImage new_image = new BufferedImage(i * getBlockSize(), h * getBlockSize(), BufferedImage.TYPE_INT_ARGB);
        Graphics2D graphics = new_image.createGraphics();
        for (int j = 0; j < h; j++) {
          create_image_row(j, i, image, graphics);
        }
        graphics.dispose();
        setBackground(new Sprite(new_image));


        try {
            stream.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    public void create_image_row(int row_number, int width, BufferedImage image, Graphics2D graphics){
        for (int i = 0; i < width; i++) {
            int pixel = image.getRGB(i, row_number);
            BufferedImage g_image = image_for(pixel, image);
            graphics.drawImage(g_image, null, i* getBlockSize(), row_number * getBlockSize());
        }
    }

    public BufferedImage image_for(int color, BufferedImage image){
        BufferedImage newImage = new BufferedImage(getBlockSize(), getBlockSize(), BufferedImage.TYPE_3BYTE_BGR);
        Graphics graphics = newImage.createGraphics();
        Color kolor = colorMap.get(color);
        graphics.setColor(kolor);
        graphics.fillRect(0,0,getBlockSize(), getBlockSize());
        graphics.dispose();
        return newImage;
    }

    public int getBlockSize() {
        return blockSize;
    }

    public void setBlockSize(int blockSize) {
        this.blockSize = blockSize;
    }

    public List<List<Boolean>> getWalkingMatrix() {
        return walkingMatrix;
    }

    public void setWalkingMatrix(List<List<Boolean>> walkingMatrix) {
        this.walkingMatrix = walkingMatrix;
    }

    public Sprite getBackground() {
        return background;
    }

    public void setBackground(Sprite background) {
        this.background = background;
    }
}
